# Sales Data — Inspection and Quality Checks

This notebook loads the raw sales dataset and examines its structure, data types, missing values, duplicates, and obvious data-quality issues.

**Important:** No cleaning or modification is performed in this notebook.

## 1. Import Libraries

`Path` manages file paths and `pandas` provides the DataFrame tools used for inspection.

In [1]:
from pathlib import Path

import pandas as pd

## 2. Locate the Dataset

The path logic works when the notebook starts from either the project root or the `notebooks` directory.

In [2]:
PROJECT_ROOT = Path.cwd()

if not (PROJECT_ROOT / "data").exists():
    PROJECT_ROOT = PROJECT_ROOT.parent

DATA_PATH = PROJECT_ROOT / "data" / "raw" / "sales_data.xlsx"

print("Project root:", PROJECT_ROOT)
print("Dataset path:", DATA_PATH)
print("Dataset exists:", DATA_PATH.exists())

Project root: c:\Users\Victus 16\sales-data-analysis
Dataset path: c:\Users\Victus 16\sales-data-analysis\data\raw\sales_data.xlsx
Dataset exists: True


## 3. Load the Excel File

`pd.read_excel()` reads the worksheet and returns a Pandas DataFrame.

In [3]:
df = pd.read_excel(
    DATA_PATH,
    sheet_name="Raw_Sales_Data",
)

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 4. Preview the Data

In [4]:
df.head()

,Order_Line_ID,Order_ID,Order_Date,Customer_ID,Product,Category,City,Quantity,Unit_Price_USD,Discount_Rate,Payment_Method
0,LINE-00001,ORD-2025-0001,2025-11-24 00:00:00,CUS-0058,Keyboard,Accessories,New York,3,77.12,0.15,PayPal
1,LINE-00002,ORD-2025-0002,2025-02-14 00:00:00,CUS-0303,Wireless Mouse,Accessories,Seattle,1,45.73,0.05,Credit Card
2,LINE-00003,ORD-2025-0003,2025-11-29 00:00:00,CUS-0280,USB-C Hub,Accessories,NaN,3,59.87,0.10,Debit Card
3,LINE-00004,ORD-2025-0004,2025-12-24 00:00:00,CUS-0217,Webcam,Accessories,Austin,1,88.27,0.00,PayPal
4,LINE-00005,ORD-2025-0005,2025-06-26 00:00:00,CUS-0310,External SSD,Storage,Austin,5,147.08,0.05,Credit Card


In [5]:
df.tail()

,Order_Line_ID,Order_ID,Order_Date,Customer_ID,Product,Category,City,Quantity,Unit_Price_USD,Discount_Rate,Payment_Method
1290,LINE-01133,ORD-2025-0704,2025-10-21 00:00:00,CUS-0159,Headphones,Audio,Boston,1,122.48,0.05,Bank Transfer
1291,LINE-00058,ORD-2025-0036,2025-02-05 00:00:00,CUS-0143,External SSD,Storage,Austin,4,140.77,0.20,Bank Transfer
1292,LINE-00814,ORD-2025-0508,2025-01-20 00:00:00,CUS-0104,Bluetooth Speaker,Audio,New York,1,107.19,0.05,Bank Transfer
1293,LINE-00199,ORD-2025-0132,2025-05-07 00:00:00,CUS-0225,USB-C Hub,Accessories,Seattle,1,67.85,0.05,Debit Card
1294,LINE-00240,ORD-2025-0158,2025-10-15 00:00:00,CUS-0061,Wireless Mouse,Accessories,Austin,1,48.00,0.05,PayPal


## 5. Inspect the Dataset Structure

In [6]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])
print("Unique orders:", df["Order_ID"].nunique())

Rows: 1295
Columns: 11
Unique orders: 800


In [7]:
df.columns.tolist()

['Order_Line_ID',
 'Order_ID',
 'Order_Date',
 'Customer_ID',
 'Product',
 'Category',
 'City',
 'Quantity',
 'Unit_Price_USD',
 'Discount_Rate',
 'Payment_Method']

In [8]:
df.dtypes

Order_Line_ID         str
Order_ID              str
Order_Date         object
Customer_ID           str
Product               str
Category              str
City                  str
Quantity           object
Unit_Price_USD    float64
Discount_Rate     float64
Payment_Method        str
dtype: object

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1295 entries, 0 to 1294
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Order_Line_ID   1295 non-null   str    
 1   Order_ID        1295 non-null   str    
 2   Order_Date      1295 non-null   object 
 3   Customer_ID     1295 non-null   str    
 4   Product         1295 non-null   str    
 5   Category        1295 non-null   str    
 6   City            1285 non-null   str    
 7   Quantity        1295 non-null   object 
 8   Unit_Price_USD  1295 non-null   float64
 9   Discount_Rate   1295 non-null   float64
 10  Payment_Method  1287 non-null   str    
dtypes: float64(2), object(2), str(7)
memory usage: 111.4+ KB


## 6. Descriptive Statistics

`describe()` summarizes columns that Pandas currently recognizes as numeric.

In [10]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Unit_Price_USD,1295.0,171.438432,225.555017,-100.0,56.38,94.26,158.695,1021.4
Discount_Rate,1295.0,0.054208,0.093385,-0.1,0.00,0.05,0.100,2.0


## 7. Missing Values

In [11]:
missing_values = df.isna().sum()
missing_values[missing_values > 0]

City              10
Payment_Method     8
dtype: int64

## 8. Exact Duplicate Rows

In [12]:
duplicate_count = df.duplicated().sum()
print("Exact duplicate rows:", duplicate_count)

Exact duplicate rows: 8


In [13]:
df[df.duplicated(keep=False)].sort_values("Order_Line_ID")

,Order_Line_ID,Order_ID,Order_Date,Customer_ID,Product,Category,City,Quantity,Unit_Price_USD,Discount_Rate,Payment_Method
57,LINE-00058,ORD-2025-0036,2025-02-05 00:00:00,CUS-0143,External SSD,Storage,Austin,4,140.77,0.20,Bank Transfer
1291,LINE-00058,ORD-2025-0036,2025-02-05 00:00:00,CUS-0143,External SSD,Storage,Austin,4,140.77,0.20,Bank Transfer
198,LINE-00199,ORD-2025-0132,2025-05-07 00:00:00,CUS-0225,USB-C Hub,Accessories,Seattle,1,67.85,0.05,Debit Card
1293,LINE-00199,ORD-2025-0132,2025-05-07 00:00:00,CUS-0225,USB-C Hub,Accessories,Seattle,1,67.85,0.05,Debit Card
239,LINE-00240,ORD-2025-0158,2025-10-15 00:00:00,CUS-0061,Wireless Mouse,Accessories,Austin,1,48.00,0.05,PayPal
1294,LINE-00240,ORD-2025-0158,2025-10-15 00:00:00,CUS-0061,Wireless Mouse,Accessories,Austin,1,48.00,0.05,PayPal
321,LINE-00322,ORD-2025-0209,2025-04-08 00:00:00,CUS-0027,USB Flash Drive,Storage,Austin,4,28.50,0.10,Bank Transfer
1287,LINE-00322,ORD-2025-0209,2025-04-08 00:00:00,CUS-0027,USB Flash Drive,Storage,Austin,4,28.50,0.10,Bank Transfer
735,LINE-00736,ORD-2025-0463,2025-10-11 00:00:00,CUS-0262,Bluetooth Speaker,Audio,Boston,1,103.64,0.00,Debit Card
1289,LINE-00736,ORD-2025-0463,2025-10-11 00:00:00,CUS-0262,Bluetooth Speaker,Audio,Boston,1,103.64,0.00,Debit Card


## 9. Inspect City and Quantity Values

In [14]:
df["City"].value_counts(dropna=False)

City
Austin              228
San Francisco       221
Chicago             217
Seattle             215
Boston              202
New York            186
NaN                  10
  San Francisco       3
  Boston              2
boston                2
SAN FRANCISCO         2
  Chicago             1
chicago               1
san francisco         1
  Austin              1
CHICAGO               1
  Seattle             1
BOSTON                1
Name: count, dtype: int64

In [15]:
df["Quantity"].value_counts(dropna=False)

Quantity
1          520
2          348
3          233
4          118
5           70
0            2
unknown      1
-1           1
-2           1
three        1
Name: count, dtype: int64

## 10. Inspect Invalid Numeric Values

In [16]:
invalid_prices = df[df["Unit_Price_USD"] <= 0][
    ["Order_Line_ID", "Product", "Unit_Price_USD"]
]

print("Invalid unit prices:", len(invalid_prices))
invalid_prices

Invalid unit prices: 5


,Order_Line_ID,Product,Unit_Price_USD
517,LINE-00518,Desk Lamp,0.0
619,LINE-00620,Bluetooth Speaker,0.0
767,LINE-00768,Headphones,-25.0
871,LINE-00872,Headphones,0.0
1160,LINE-01161,Wireless Mouse,-100.0


In [17]:
invalid_discounts = df[
    ~df["Discount_Rate"].between(0, 1)
][
    ["Order_Line_ID", "Product", "Discount_Rate"]
]

print("Invalid discount rates:", len(invalid_discounts))
invalid_discounts

Invalid discount rates: 5


,Order_Line_ID,Product,Discount_Rate
69,LINE-00070,Bluetooth Speaker,1.20
439,LINE-00440,External SSD,2.00
784,LINE-00785,Monitor,-0.05
847,LINE-00848,Desk Lamp,-0.10
1024,LINE-01025,Keyboard,1.50


## Initial Findings

- The dataset contains 1,295 rows and 11 columns.
- There are 800 unique orders.
- Eight exact duplicate rows exist.
- Some city and payment-method values are missing.
- City names contain whitespace and capitalization inconsistencies.
- Quantity contains invalid numeric and text values.
- Some unit prices are zero or negative.
- Some discount rates are outside the valid range.
- Some order dates are invalid or outside 2025.
- Some products are assigned to incorrect categories.
- No data was cleaned or modified in this notebook.

## 11. Formal Data-Quality Checks

This section validates the dataset against the rules defined in the data dictionary. No source data is modified.

### 11.1 Validate the Column Structure

In [18]:
EXPECTED_COLUMNS = [
    "Order_Line_ID",
    "Order_ID",
    "Order_Date",
    "Customer_ID",
    "Product",
    "Category",
    "City",
    "Quantity",
    "Unit_Price_USD",
    "Discount_Rate",
    "Payment_Method",
]

missing_columns = [
    column
    for column in EXPECTED_COLUMNS
    if column not in df.columns
]

unexpected_columns = [
    column
    for column in df.columns
    if column not in EXPECTED_COLUMNS
]

print("Missing columns:", missing_columns)
print("Unexpected columns:", unexpected_columns)

Missing columns: []
Unexpected columns: []


### 11.2 Define Valid Reference Values

In [19]:
VALID_CITIES = [
    "New York",
    "Chicago",
    "Austin",
    "Seattle",
    "Boston",
    "San Francisco",
]

PRODUCT_CATEGORY_MAP = {
    "Laptop": "Computers",
    "Monitor": "Computers",
    "Keyboard": "Accessories",
    "Wireless Mouse": "Accessories",
    "USB-C Hub": "Accessories",
    "Webcam": "Accessories",
    "Headphones": "Audio",
    "Bluetooth Speaker": "Audio",
    "External SSD": "Storage",
    "USB Flash Drive": "Storage",
    "Office Chair": "Furniture",
    "Desk Lamp": "Furniture",
}

### 11.3 Validate Quantity

In [20]:
quantity_numeric = pd.to_numeric(
    df["Quantity"],
    errors="coerce",
)

non_numeric_quantity_count = quantity_numeric.isna().sum()
non_positive_quantity_count = (quantity_numeric <= 0).sum()

print("Non-numeric quantities:", non_numeric_quantity_count)
print("Zero or negative quantities:", non_positive_quantity_count)

Non-numeric quantities: 2
Zero or negative quantities: 4


### 11.4 Validate Order Dates

In [21]:
parsed_dates = pd.to_datetime(
    df["Order_Date"],
    errors="coerce",
)

unparseable_date_count = parsed_dates.isna().sum()
outside_2025_count = (
    parsed_dates.notna()
    & (parsed_dates.dt.year != 2025)
).sum()

print("Unparseable dates:", unparseable_date_count)
print("Dates outside 2025:", outside_2025_count)

Unparseable dates: 4
Dates outside 2025: 1


### 11.5 Validate City Names

In [22]:
non_null_cities = df["City"].dropna()
non_standard_city_count = (
    ~non_null_cities.isin(VALID_CITIES)
).sum()

print("Non-standard city values:", non_standard_city_count)

Non-standard city values: 16


### 11.6 Validate Product and Category Relationships

In [23]:
expected_categories = df["Product"].map(
    PRODUCT_CATEGORY_MAP
)

category_mismatch_count = (
    df["Category"] != expected_categories
).sum()

print("Product/category mismatches:", category_mismatch_count)

Product/category mismatches: 5


### 11.7 Build the Quality Summary Table

Each count represents a separate validation rule. Counts should not be added together because one row may fail multiple checks.

In [24]:
quality_summary = pd.DataFrame(
    {
        "Quality_Check": [
            "Exact duplicate rows",
            "Duplicate Order_Line_ID values",
            "Missing city values",
            "Missing payment methods",
            "Non-standard city values",
            "Non-numeric quantities",
            "Zero or negative quantities",
            "Zero or negative unit prices",
            "Discount rates outside 0-1",
            "Unparseable order dates",
            "Dates outside 2025",
            "Product/category mismatches",
        ],
        "Issue_Count": [
            df.duplicated().sum(),
            df["Order_Line_ID"].duplicated().sum(),
            df["City"].isna().sum(),
            df["Payment_Method"].isna().sum(),
            non_standard_city_count,
            non_numeric_quantity_count,
            non_positive_quantity_count,
            (df["Unit_Price_USD"] <= 0).sum(),
            (~df["Discount_Rate"].between(0, 1)).sum(),
            unparseable_date_count,
            outside_2025_count,
            category_mismatch_count,
        ],
    }
)

quality_summary

,Quality_Check,Issue_Count
0,Exact duplicate rows,8
1,Duplicate Order_Line_ID values,8
2,Missing city values,10
3,Missing payment methods,8
4,Non-standard city values,16
5,Non-numeric quantities,2
6,Zero or negative quantities,4
7,Zero or negative unit prices,5
8,Discount rates outside 0-1,5
9,Unparseable order dates,4
